tests/testthat/test-stat-density.R

```diff
@@ -1,3 +1,54 @@
+test_that("stat_density actually computes density", {
+  # Compare functon approximations because outputs from `ggplot()` and
+  # `density()` give grids spanning different ranges
+  dens <- stats::density(mtcars$mpg)
+  expected_density_fun <- stats::approxfun(data.frame(x = dens$x, y = dens$y))
+
+  plot <- ggplot(mtcars, aes(mpg)) + stat_density()
+  actual_density_fun <- stats::approxfun(layer_data(plot)[, c("x", "y")])
+
+  test_sample <- unique(mtcars$mpg)
+  expect_equal(
+    expected_density_fun(test_sample),
+    actual_density_fun(test_sample),
+    tolerance = 1e-3
+  )
+})
+
+test_that("stat_density uses `bounds`", {
+  make_density <- function(bounds) {
+    plot <- ggplot(mtcars, aes(mpg)) + stat_density(bounds = bounds)
+    return(layer_data(plot)[, c("x", "y")])
+  }
+
+  mpg_min <- min(mtcars$mpg)
+  mpg_max <- max(mtcars$mpg)
+
+  # Density should be around twice higher at finite boundary
+  expect_snapshot(make_density(c(-Inf, Inf)))
+  expect_snapshot(make_density(c(mpg_min, Inf)))
+  expect_snapshot(make_density(c(-Inf, mpg_max)))
+  expect_snapshot(make_density(c(mpg_min, mpg_max)))
```

Comment:
Is there a reason that you're using snapshots here? Couldn't you just test numerically that density is approximately twice higher at finite boundary? It doesn't have to be that accurate. You could just test that it's > 1.9x and < 2.1x, for example.

Reply:
Snapshots seem to provide more full testing of the feature. This is not only about being twice as higher at finite boundary, but about a specific way of dealing with boundary correction (via density reflection). I can make snapshots smaller by setting smaller `n`.

All in all, I'd keep them. Will do as you ask, of course. So should I not use snapshots?

Reply:
I don't like snapshots because if they break at some point in the future nobody will know what the correct result should have been. If you want to test the full reflection, then the right way to do this is to calculate the reflection again in the test and compare numerically. Yes, it may seem silly to run the same calculation in two locations and compare they're the same, but it actually is a good check for the correctness of the implementation in the main library.

Reply:
Updated test to not use snapshots.